# vim: se ft=diff :
# HG changeset patch
# User Yonggang Luo <luoyonggang@gmail.com>
# Date 1630669259 -7200
# Parent  a2d966049c49b14bd6628edc1b6bd899b978d680
Add StringizeProperties(const AtomArray& aPropertiesArray) in nsTreeUtils

diff --git a/layout/xul/tree/nsTreeUtils.cpp b/layout/xul/tree/nsTreeUtils.cpp
--- a/layout/xul/tree/nsTreeUtils.cpp
+++ b/layout/xul/tree/nsTreeUtils.cpp
@@ -42,16 +42,28 @@ nsresult nsTreeUtils::TokenizeProperties
 
     RefPtr<nsAtom> atom = NS_Atomize(Substring(first, iter));
     aPropertiesArray.AppendElement(atom);
   } while (iter != end);
 
   return NS_OK;
 }
 
+nsAutoString nsTreeUtils::StringizeProperties(
+    const AtomArray& aPropertiesArray) {
+  nsAutoString properties;
+  for (uint32_t i = 0; i < aPropertiesArray.Length(); ++i) {
+    if (i > 0) {
+      properties.Append(' ');
+    }
+    properties += aPropertiesArray[i]->GetUTF16String();
+  }
+  return properties;
+}
+
 nsIContent* nsTreeUtils::GetImmediateChild(nsIContent* aContainer,
                                            nsAtom* aTag) {
   dom::FlattenedChildIterator iter(aContainer);
   for (nsIContent* child = iter.GetNextChild(); child;
        child = iter.GetNextChild()) {
     if (child->IsXULElement(aTag)) {
       return child;
     }
diff --git a/layout/xul/tree/nsTreeUtils.h b/layout/xul/tree/nsTreeUtils.h
--- a/layout/xul/tree/nsTreeUtils.h
+++ b/layout/xul/tree/nsTreeUtils.h
@@ -24,16 +24,18 @@ class nsTreeUtils {
  public:
   /**
    * Parse a whitespace separated list of properties into an array
    * of atoms.
    */
   static nsresult TokenizeProperties(const nsAString& aProperties,
                                      mozilla::AtomArray& aPropertiesArray);
 
+  static nsAutoString StringizeProperties(const mozilla::AtomArray& aPropertiesArray);
+
   static nsIContent* GetImmediateChild(nsIContent* aContainer, nsAtom* aTag);
 
   static nsIContent* GetDescendantChild(nsIContent* aContainer, nsAtom* aTag);
 
   static nsresult UpdateSortIndicators(mozilla::dom::Element* aColumn,
                                        const nsAString& aDirection);
 
   static nsresult GetColumnIndex(mozilla::dom::Element* aColumn,
